### DISCUSSION TITLE
How can I change PATH modification between user and system with perUserOrMachine?

### DISCUSSION THREAD

**Discussion by davidanthoff** [wixtoolset/issues#7578] [Questions] [answered]
I'm currently working on a new MSI installer for the [Julia](https://julialang.org/) language, and am stuck on (hopefully) one last thing: I use `Scope="perUserOrMachine"` and that seems to work, except for the modification of the `PATH` environment variable in the installer. Currently I have an element
```
<Environment Id='UpdatePath' Name='PATH' Action='set' Permanent='no' System='yes' Part='last' Value='[INSTALLFOLDER]' />
```
in there, but the `System='yes'` of course doesn't work correctly when things are installed per user. Essentially I need to find some way to have the value of `System` be `yes` for a machine install and `no` for a user install. Any help on how I might be able to do that would be great :)

The entire wxs file as it stands right now is [here](https://github.com/JuliaLang/juliaup/blob/d3c894402ea1091b9d86cdb7e865d37f71ba5a4e/deploy/msi/Julia.wxs) if that is helpful.

Thanks!

**Comment by davidanthoff** [comment]
I should say, it would also be super helpful if someone could point me to an example of an existing installer that does this. This entire scenario doesn't strike me as very niche, I would think others must have solved this somehow?

**Reply by robmen** [reply]
`perUserOrMachine` switching is not very common. It can introduce complexity as the feature was added in later Windows Installer version, I think MSI v4.5, thus not baked in since the beginning.

You'll probably need to do a bit more switching on your Value to handle setting the Prefix yourself. See the [Environment Table documentation](https://learn.microsoft.com/en-us/windows/win32/msi/environment-table) for the nitty gritty details.

**Reply by davidanthoff** [reply, marked as answer]
Ah, that link was very helpful, thanks!

My takeaway from that is that I won't be able to handle this with one entry in the `Environment` table. Whether something is installed system or user depends on the prefix of the _name_ in that table, and the type of that is `Text`, which I assume means I can't make that depend on the value of some property during install.

What I ended up doing now is just have two components with a condition `ALLUSERS=2 OR MSIINSTALLPERUSER=1` for one and `NOT (ALLUSERS=2 OR MSIINSTALLPERUSER=1)` for the other, and one has an environment element with `System='yes'` and the other one with `System='no'`. That part looks like [this](https://github.com/JuliaLang/juliaup/blob/83dee19db90d7137abb2c97c4a98a56bc27c2e62/deploy/msi/Julia.wxs#L45-L59) now:

```
<Component Id="JuliaupExecutableSystem" Guid="2a881338-e0f2-4df7-8c02-71373891961c" Condition="NOT (ALLUSERS=2 OR MSIINSTALLPERUSER=1)">
    <File Id="JuliaupEXESystem" Name="juliaup.exe" DiskId="1" Source="juliaup.exe" KeyPath="yes" />
    <Environment Id='UpdatePathSystem' Name='PATH' Action='set' Permanent='no' System='yes' Part='last' Value='[INSTALLFOLDER]' />
</Component>

<Component Id="JuliaupExecutableUser" Guid="587bc90d-fcc9-4163-8e90-342a0f4178cb" Condition="ALLUSERS=2 OR MSIINSTALLPERUSER=1">
    <File Id="JuliaupEXEUser" Name="juliaup.exe" DiskId="1" Source="juliaup.exe" KeyPath="yes" />
    <Environment Id='UpdatePathUser' Name='PATH' Action='set' Permanent='no' System='no' Part='last' Value='[INSTALLFOLDER]' />
</Component>
```
At least from some very quick testing that seems to work. @robmen if you see anything with that strategy that worries you, I'd appreciate a warning. But in any case, thanks so much for the pointer!

The entire installer now is https://github.com/JuliaLang/juliaup/blob/83dee19db90d7137abb2c97c4a98a56bc27c2e62/deploy/msi/Julia.wxs, in case anyone wants to see this. If anyone spots any other problems with that, please also let me know!

**Reply by robmen** [reply]
That probably works. I think there is a simpler solution using the Value with the right prefix but I'd have to play with it. Maybe something I'd try on the Deployment Dojo some time.

**Reply by davidanthoff** [reply]
I think the problem though is that the prefix that controls user vs system only works in the `Name`, not the `Value` column? At least that is how I read the documentation.
